File: go-backend-interview-prep/10_design_patterns_in_go/11_hexagonal_ports_adapters_mini/main.py
"""Hexagonal architecture (ports & adapters) -- Python equivalent.

Core depends on duck-typed ports. Adapters implement ports.
main() wires adapters to core.
"""

from dataclasses import dataclass, field
from datetime import datetime


# ============================================================
# CORE (depends on ports only -- duck typing in Python)
# ============================================================

@dataclass
class Notification:
    id: str
    to: str
    message: str
    sent_at: datetime = field(default_factory=datetime.now)


class NotificationService:
    """Core service -- depends on sender_port and repo_port (duck-typed)."""

    def __init__(self, sender, repo):
        self.sender = sender  # SenderPort
        self.repo = repo      # RepoPort

    def send(self, id: str, to: str, message: str) -> None:
        if not to:
            raise ValueError("recipient required")
        self.sender.send(to, message)
        n = Notification(id=id, to=to, message=message)
        self.repo.save(n)


# ============================================================
# ADAPTERS (implement ports)
# ============================================================

class EmailAdapter:
    """Adapter for SenderPort."""
    def send(self, to: str, message: str) -> None:
        print(f"  [EMAIL ADAPTER] Sent to {to}: {message}")


class MemoryRepoAdapter:
    """Adapter for RepoPort."""
    def __init__(self):
        self._data: dict[str, Notification] = {}

    def save(self, n: Notification) -> None:
        self._data[n.id] = n
        print(f"  [REPO ADAPTER]  Saved notification {n.id}")

    def find_by_id(self, id: str) -> Notification:
        if id not in self._data:
            raise KeyError(f"not found: {id}")
        return self._data[id]


# --- Fakes for testing ---

class FakeSender:
    def __init__(self):
        self.calls: list[str] = []

    def send(self, to: str, message: str) -> None:
        self.calls.append(f"{to}:{message}")


class FakeRepo:
    def __init__(self):
        self.saved: list[Notification] = []

    def save(self, n: Notification) -> None:
        self.saved.append(n)

    def find_by_id(self, id: str) -> Notification:
        for n in self.saved:
            if n.id == id:
                return n
        raise KeyError(f"not found: {id}")


# ============================================================
# MAIN (wires adapters to ports)
# ============================================================

def main() -> None:
    print("=== Hexagonal Architecture (Ports & Adapters) ===")
    print()

    # Production wiring
    print("--- Production: email + memory repo ---")
    svc = NotificationService(EmailAdapter(), MemoryRepoAdapter())
    svc.send("n1", "alice@example.com", "Order confirmed")
    svc.send("n2", "bob@example.com", "Payment received")
    print()

    # Test wiring: fakes
    print("--- Test: fake sender + fake repo ---")
    fake_sender = FakeSender()
    fake_repo = FakeRepo()
    test_svc = NotificationService(fake_sender, fake_repo)
    test_svc.send("t1", "test@test.com", "test message")

    print(f"  Sender calls: {fake_sender.calls}")
    print(f"  Repo saved:   {len(fake_repo.saved)} notification(s)")
    print()

    print("Architecture:")
    print("  Core:     NotificationService (depends on sender + repo duck types)")
    print("  Adapters: EmailAdapter, MemoryRepoAdapter (implement duck types)")
    print("  main():   Wires adapters to core (dependency injection)")
    print("  Tests:    Swap adapters for fakes -- zero core changes")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/10_design_patterns_in_go/11_hexagonal_ports_adapters_mini/main.go
package main

import (
	"errors"
	"fmt"
	"time"
)

// ============================================================
// CORE (depends on ports only -- zero external imports)
// ============================================================

// --- Domain ---

type Notification struct {
	ID      string
	To      string
	Message string
	SentAt  time.Time
}

// --- Ports (interfaces defined by core) ---

type SenderPort interface {
	Send(to, message string) error
}

type RepoPort interface {
	Save(n Notification) error
	FindByID(id string) (Notification, error)
}

// --- Core service (depends on ports, not adapters) ---

type NotificationService struct {
	sender SenderPort
	repo   RepoPort
}

func NewNotificationService(sender SenderPort, repo RepoPort) *NotificationService {
	return &NotificationService{sender: sender, repo: repo}
}

func (s *NotificationService) Send(id, to, message string) error {
	// Business logic lives here -- no knowledge of email/DB specifics
	if to == "" {
		return errors.New("recipient required")
	}
	if err := s.sender.Send(to, message); err != nil {
		return fmt.Errorf("send failed: %w", err)
	}
	n := Notification{ID: id, To: to, Message: message, SentAt: time.Now()}
	return s.repo.Save(n)
}

// ============================================================
// ADAPTERS (implement ports -- live outside core)
// ============================================================

// --- Email adapter (implements SenderPort) ---

type EmailAdapter struct{}

func (e *EmailAdapter) Send(to, message string) error {
	fmt.Printf("  [EMAIL ADAPTER] Sent to %s: %s\n", to, message)
	return nil
}

// --- In-memory repo adapter (implements RepoPort) ---

type MemoryRepoAdapter struct {
	data map[string]Notification
}

func NewMemoryRepoAdapter() *MemoryRepoAdapter {
	return &MemoryRepoAdapter{data: make(map[string]Notification)}
}

func (r *MemoryRepoAdapter) Save(n Notification) error {
	r.data[n.ID] = n
	fmt.Printf("  [REPO ADAPTER]  Saved notification %s\n", n.ID)
	return nil
}

func (r *MemoryRepoAdapter) FindByID(id string) (Notification, error) {
	n, ok := r.data[id]
	if !ok {
		return Notification{}, errors.New("not found: " + id)
	}
	return n, nil
}

// --- Fake adapters for testing ---

type FakeSender struct{ Calls []string }

func (f *FakeSender) Send(to, msg string) error {
	f.Calls = append(f.Calls, to+":"+msg)
	return nil
}

type FakeRepo struct{ Saved []Notification }

func (f *FakeRepo) Save(n Notification) error {
	f.Saved = append(f.Saved, n)
	return nil
}
func (f *FakeRepo) FindByID(id string) (Notification, error) {
	for _, n := range f.Saved {
		if n.ID == id {
			return n, nil
		}
	}
	return Notification{}, errors.New("not found")
}

// ============================================================
// MAIN (wires adapters to ports)
// ============================================================

func main() {
	fmt.Println("=== Hexagonal Architecture (Ports & Adapters) ===")
	fmt.Println()

	// --- Production wiring ---
	fmt.Println("--- Production: email + memory repo ---")
	svc := NewNotificationService(
		&EmailAdapter{},        // adapter for SenderPort
		NewMemoryRepoAdapter(), // adapter for RepoPort
	)
	svc.Send("n1", "alice@example.com", "Order confirmed")
	svc.Send("n2", "bob@example.com", "Payment received")
	fmt.Println()

	// --- Test wiring: fakes ---
	fmt.Println("--- Test: fake sender + fake repo ---")
	fakeSender := &FakeSender{}
	fakeRepo := &FakeRepo{}
	testSvc := NewNotificationService(fakeSender, fakeRepo)
	testSvc.Send("t1", "test@test.com", "test message")

	fmt.Printf("  Sender calls: %v\n", fakeSender.Calls)
	fmt.Printf("  Repo saved:   %d notification(s)\n", len(fakeRepo.Saved))
	fmt.Println()

	fmt.Println("Architecture:")
	fmt.Println("  Core:     NotificationService (depends on SenderPort + RepoPort)")
	fmt.Println("  Adapters: EmailAdapter, MemoryRepoAdapter (implement ports)")
	fmt.Println("  main():   Wires adapters to ports (dependency injection)")
	fmt.Println("  Tests:    Swap adapters for fakes -- zero core changes")
}
